bundles/binding/org.openhab.binding.smarthomatic/src/main/java/org/openhab/binding/smarthomatic/internal/BaseStation.java

```diff
@@ -0,0 +1,279 @@
+/**
+ * Copyright (c) 2010-2015, openHAB.org and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.smarthomatic.internal;
+
+import java.util.StringTokenizer;
+
+import org.openhab.core.library.types.DecimalType;
+import org.openhab.core.library.types.HSBType;
+import org.openhab.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * these class represents the basestation and the communication between openhab
+ * and the smarthomatic basestation
+ * {@link https://www.smarthomatic.org/devices/base_station.html}
+ * 
+ * @author mcjobo
+ * @author arohde
+ * @since 1.8.0
+ */
+public class BaseStation implements SerialEventWorker {
+
+	private static final Logger logger = LoggerFactory.getLogger(BaseStation.class);
+
+	private SerialDevice serialDevice;
+
+	private String[] versionInfo;
+
+	private SerialEventWorker bindingEventWorker;
+
+	/* always 0 for base station */
+	// private final int deviceType = 0;
+	/* can be 0..4095, normally 0 */
+	// private int deviceId = 0;
+
+	/*
+	 * supported items are: StringItem SwitchItem
+	 */
+
+	/**
+	 * public constructor for BaseStation
+	 * 
+	 * @since 1.8.0
+	 * @param port
+	 *            of the serial line uses eg com4 or /dev/ttyUSB0
+	 * @param baud
+	 *            rate used eg 19200
+	 * @param sew
+	 *            used event worker
+	 */
+	public BaseStation(String port, int baud, SerialEventWorker sew) {
+		bindingEventWorker = sew;
+		serialDevice = new SerialDevice(port, baud);
+		serialDevice.setEventWorker(this);
+		try {
+			serialDevice.initialize();
+		} catch (InitializationException e) {
+			logger.error("Port cannot be opened. Port: {}", port);
+			logger.error("initialization exception", e);
+		}
+	}
+
+	/**
+	 * closes the serial port
+	 * 
+	 * @since 1.8.0
+	 */
+	public void closeSerialPort() {
+		serialDevice.close();
+	}
+
+	private String genHexString(int deviceID, int length) {
+		String s = Integer.toHexString(deviceID);
+		int len = s.length();
+		for (int i = len; i < length; i++) {
+			s = "0" + s;
+		}
+		return s;
+	}
+
+	private String getToggleTime(int toggleTime, boolean on) {
+		// shift left the on bit, so that the on-bit stands on position 17 from
+		// right
+		long temp = (on ? 1 : 0) << 16;
+		// patch toggletime into the last 16 bit and be sure to prevent to
+		// override on-bit
+		temp |= (toggleTime & 0x0001FFFF);
+		// shift temp to left by 15 bits (the on bit should stand at msb
+		// position)
+		temp <<= 15;
+
+		// write hex representation of temp to s
+		String s = Long.toHexString(temp);
+
+		// fill String with leading zeros until it's minimum 2 chars long
+		for (int i = s.length(); i < 2; i++)
```

Comment:
Please add braces on the for loop, even if it is one line only (see line 82 in this file)